src/main/java/org/folio/services/instance/InstanceS3Service.java

```diff
@@ -0,0 +1,246 @@
+    return vertx.executeBlocking(() -> {
+      FolioS3Client s3Client = folioS3ClientFactory.getFolioS3Client();
+      InputStream inputStream = s3Client.read(bulkRequest.getRecordsFileName());
+
+      try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
+        return reader.lines()
+          .map(JsonObject::new)
+          .toList();
+      }
+    });
+  }
+
+  private Future<InstanceBulkResponse> upsert(List<JsonObject> instances, InstanceBulkRequest bulkRequest) {
+    BulkProcessingContext bulkProcessingContext = new BulkProcessingContext(bulkRequest.getRecordsFileName());
+    List<Pair<Instance, List<PrecedingSucceedingTitle>>> instancesPairs = instances.stream()
+      .map(json -> Pair.of(mapInstanceDtoJsonToInstance(json), extractPrecedingSucceedingTitles(json)))
+      .toList();
+
+    return upsert(instancesPairs)
+      .map(v -> new InstanceBulkResponse().withErrorsNumber(0))
+      .recover(e -> processSequentially(instancesPairs, bulkProcessingContext));
+  }
+
+  private Future<Void> upsert(List<Pair<Instance, List<PrecedingSucceedingTitle>>> instances) {
+    List<Instance> instanceList = instances.stream().map(Pair::getLeft).toList();
+
+    return ensureInstancesWithNonMarcControlledFields(instanceList)
+      .compose(v -> instanceService.createInstances(instanceList, true, true))
+      .compose(response -> {
+        if (!isCreateSuccessResponse(response)) {
+          String msg = String.format("Failed to update instances, status: '%s', message: '%s'",
+            response.getStatus(), Json.encode(response.getEntity()));
+          return Future.failedFuture(msg);
+        }
+        return Future.succeededFuture();
+      })
+      .compose(v -> updatePrecedingSucceedingTitles(instances));
+  }
+
+  private List<PrecedingSucceedingTitle> extractPrecedingSucceedingTitles(JsonObject instanceJson) {
+    List<PrecedingSucceedingTitle> titles = new ArrayList<>();
+    instanceJson.getJsonArray(PRECEDING_TITLES_FIELD).stream()
+      .map(JsonObject.class::cast)
+      .map(title -> title.mapTo(PrecedingSucceedingTitle.class))
+      .map(title -> title.withSucceedingInstanceId(instanceJson.getString(ID_FIELD)))
+      .forEach(titles::add);
+
+    instanceJson.getJsonArray(SUCCEEDING_TITLES_FIELD).stream()
+      .map(JsonObject.class::cast)
+      .map(title -> title.mapTo(PrecedingSucceedingTitle.class))
+      .map(title -> title.withPrecedingInstanceId(instanceJson.getString(ID_FIELD)))
+      .forEach(titles::add);
+    return titles;
+  }
+
+  private Future<Void> ensureInstancesWithNonMarcControlledFields(List<Instance> instances) {
+    return instanceRepository.getById(instances, Instance::getId).map(existingInstances -> {
+      instances.forEach(instance -> {
+        if (existingInstances.get(instance.getId()) != null) {
+          InstanceUtil.copyNonMarcControlledFields(instance, existingInstances.get(instance.getId()));
+        }
+      });
+      return null;
```

Comment:
Perhaps it would be better if we returned Future.succeededFuture() instead of null?

Reply:
rewritten